doc/execution.rst

```diff
@@ -0,0 +1,219 @@
+itself.  In R4.3 and above, ``qrexec-daemon`` directly submits or executes the call,
+and ``qrexec-client`` is only used for calls submitted by dom0.
+
+Qubes RPC service execution
+---------------------------
+
+A Qubes RPC service is a file, socket, or symbolic link under ``/etc/qubes-rpc/``
+or ``/usr/local/etc/qubes-rpc``.  When a call is made, ``qrexec-agent`` (in a VM)
+or ``qrexec-client`` (in dom0) searches for the first entry in the following list:
+
+1. ``/usr/local/etc/qubes-rpc/SERVICE+ARG``
+2. ``/etc/qubes-rpc/SERVICE+ARG``
+3. ``/usr/local/etc/qubes-rpc/SERVICE``
+4. ``/etc/qubes-rpc/SERVICE``
+
+``SERVICE`` must be replaced by the service being invoked, and ``ARG`` by its argument.
+If ``SERVICE`` is longer than ``NAME_MAX`` (255 on Linux), the call fails.  If
+``SERVICE`` is ``NAME_MAX`` or less, but ``SERVICE+ARG`` exceeds ``NAME_MAX``,
+steps 1 and 3 are skipped.  If no argument at all is provided, the search proceeds
+as if an empty argument is passed.  This is only possible for calls made by dom0,
+as VM => VM calls insert an empty argument if no argument is provided.
+
+The search is terminated by any of the following:
+
+1. There are no more paths to check.  This causes the search to fail.
+2. ``lstat(2)`` fails, setting errno to a value other than ``ENOENT``.
+   This causes the search to fail -- other paths are *not* checked,
+   even if they could be successfully executed.
+3. ``lstat(2)`` succeeds.  The search concludes successfully.
+
+If a command is not found, qrexec pretends that it exited with status 127.
+If a command cannot be executed, qrexec pretends that it exited with status 126.
+In both cases, no data is read from stdin, and no data is written to stdout or
+stderr.  However, the actual cause of the failure is logged within the VM.
+
+Symbolic links are followed when executing a service.  However, it is
+usually a mistake to rely on this behavior for executable services, as
+``-`` is allowed to be the first character in the option argument.  This
+allows option injection attacks.  Instead, a wrapper script should be used.
+
+Types of qrexec services
+------------------------
+
+There are three different types of qrexec services.  The distinction
+between service types is mostly invisible to callers.
+
+1. Executable services.  These are files with execute permission, as
+   reported by ``euidaccess(2)``.  They are executed using ``execve()``.
+   In a VM (but not in dom0), they are executed in a proper login session.
+   On Linux, PAM is used.
+
+   By default, these will run as the user passed by dom0.  This can be overridden
+   with ``user=`` in the configuration file.  Both string usernames (between single
+   quotes) and numeric user IDs are allowed.
+
+2. Socket-based services.  These are ``AF_UNIX`` stream sockets on the filesystem.
+   Data passed via stdin will be written to the socket, and data from the socket will
+   will be written to stdout.  Nothing will be written to stderr, not even EOF.
```

Comment:
Not sending EOF is arguably a bug (but a harmless one with the current client implementation).

Reply:
What I mean is - do not document it as an intended behavior.